tools/resources/__init__.py

```diff
@@ -413,6 +413,7 @@ def add_directory(
         ".h": FileType.HEADER,
         ".hh": FileType.HEADER,
         ".hpp": FileType.HEADER,
+        ".tpp": FileType.HEADER,
```

Comment:
I remember a day when header files ended with a `.h`, and we liked it.

Reply:
Do you ? Headers of the C++ library, a 35 years old language, don't have extensions. In C, X macros and friends have been traditionally put in files that don't end with `.h` too. 
If the `.h` extension as not been standardised at the language level it is because other extensions predate standardisation and these use cases are legitimate.

Reply:
Counter-pedant alert, and primed for counter-counter-pedantry.

Technically, the C++ standard doesn't say anything about file extensions.  The standard headers are included via `#include <atomic>` or whatever in the C++ source, but they could perfectly well be stored on disc as `atomic.hh` or `iso_atomic.hpp` or just built in to the compiler. Mechanism could be totally separate to whatever the `#include "xxx"` does for user headers. (Similarly `#include "myheader.h"` might not actually find the file `myheader.h`. On RISC OS it would find `h.myheader`, with `.` being the directory separator).

I actually find it a bit annoying that vendors chose to drop suffixes from the C++ `std` filenames. Having no extension at all is quite unhelpful for most other tools.

Anyway, no strong feelings about the separate prefix here. I'm also prefer to have special prefixes for things that you only want `#include`d with proper care, but it's up to tools folks to object. This one's a bit borderline, because nothing would actually break if randomly included, right? Would just slow down the build.

Reply:
@pan- There's PR to add `.inc` files, wouldnt this extension cover the use case for `.tpp` ?

Reply:
> This one's a bit borderline, because nothing would actually break if randomly included, right?

@kjbracey-arm There's no header guards `.tpp` files at the moment. With header guards nothing would break.

@0xc0170 I choose `tpp` as an extension because it is more descriptive than `h` or `hpp` (and avoid clashes with actual header files). Any extension can be use. What is expected in `inc` files ?

Reply:
I don't think we documented this - only extensions and purpose, without "what should be there". They are used currently for some autogenerated headers in TFM code.

It might be better if this extension addition is sent via separate PR where we can discuss this and alos update docs (see https://github.com/ARMmbed/mbed-os/pull/9715)

Reply:
From a tools perspective, I have no preference on the extensions. We can adapt to what you need.

The `.inc` extension is being added because the TF-M work is importing an external project into Mbed OS and the external project use `.inc` files. So we're simplifying the import process for that team.

Reply:
Does *.tpp work with other builders ie: MDK, DS-5, IAR, etc... or will this break our exporters? How is wrapping this in pragma once or ifdef... not good enough to keep the code base consistent?

Reply:
I haven't tried for all exporters (but I can!). In this case, `*.tpp` files embed template class implementations and these are instantiated only once in a cpp file dedicated to the instantiation of the implementation type that contains all the required implementations. 

These files can't be named `.cpp` as these are not compilation unit and cannot used the sole suffix `.h` as there already is a header containing the declaration so we would end up with two headers with the same name. 

Suffix for template implementation really is just a convention as long as it respects the two requirements mentioned above. I started out with the suffix `.impl.h` and am happy to change the suffix to whatever is decided. 

----

Note that it would be possible to include everything in header files (I tried) at the expense of weird include patterns, longer compilation time and random breakage requiring reordering if things get moved around. So instead I tried to keep template implementation files consistent with the way we write source implementation files (.cpp) beside the name.

Reply:
Thanks @pan- it would be good to confirm the exporters are working as well as extend the tests for exporters to exercise the file type support to match what the os is capable of building. @bridadan